crates/sui-source-validation-service/src/lib.rs

```diff
@@ -352,6 +361,35 @@ pub async fn verify_packages(config: &Config, dir: &Path) -> anyhow::Result<Netw
     Ok(lookup)
 }
 
+pub async fn watch_for_upgrades(config: &Config) -> anyhow::Result<()> {
+    let mut watch_ids = ArrayParams::new();
+    for s in &config.packages {
+        let packages = match s {
+            PackageSources::Repository(RepositorySource { packages, .. }) => packages,
+            PackageSources::Directory(DirectorySource { packages, .. }) => packages,
+        };
```

Comment:
nit (optional): Rust supports binding pattern variables over alternates in the pattern, so you can do something like this:

```suggestion
    for (
        PackageSource::Repository(RepositorySource { packages, .. }) |
        PackageSource::Directory(RepositorySource { packages, .. })
    ) in &config.packages {

```

Reply:
wow that's impressive, TIL! I think I like it, but it's maybe just a little too magic--I'm going to give it a whirl separately